src/general/world_effects/patch_events/GlobalGlaciationEvent.cs

```diff
@@ -0,0 +1,258 @@
+            }
+        }
+    }
+
+    private void TryToTriggerEvent(double totalTimePassed)
+    {
+        if (!AreConditionsMet())
+            return;
+
+        eventDuration = random.Next(Constants.GLOBAL_GLACIATION_MIN_DURATION, Constants.GLOBAL_GLACIATION_MAX_DURATION);
+        generationsLeft = eventDuration;
+
+        foreach (var (index, patch) in targetWorld.Map.Patches)
+        {
+            if (patch.Depth[0] == 0 && patch.BiomeType != BiomeType.Cave)
+            {
+                ChangePatchProperties(index, patch, totalTimePassed);
+            }
+        }
+
+        LogBeginningOfGlaciation();
+    }
+
+    private bool AreConditionsMet()
+    {
+        var numberOfSurfacePatches = 0;
+        var patchesExceedingOxygenLevel = 0;
+        foreach (var patch in targetWorld.Map.Patches.Values)
+        {
+            if (patch.Depth[0] != 0 || patch.BiomeType == BiomeType.Cave)
+                continue;
+
+            patch.Biome.TryGetCompound(Compound.Oxygen, CompoundAmountType.Biome, out var oxygenLevel);
+            if (oxygenLevel.Ambient >= Constants.GLOBAL_GLACIATION_OXYGEN_THRESHOLD)
+                patchesExceedingOxygenLevel += 1;
+
+            numberOfSurfacePatches += 1;
+        }
+
+        // Just prevent dividing by zero, but that shouldn't be possible anyway
+        numberOfSurfacePatches = numberOfSurfacePatches == 0 ? 1 : numberOfSurfacePatches;
+
+        return (float)patchesExceedingOxygenLevel / numberOfSurfacePatches >=
+            Constants.GLOBAL_GLACIATION_PATCHES_THRESHOLD
+            && random.NextFloat() <= Constants.GLOBAL_GLACIATION_CHANCE;
+    }
+
+    private void ChangePatchProperties(int index, Patch patch, double totalTimePassed)
+    {
+        modifiedPatchesIds.Add(index);
+        AdjustBackground(patch);
+        AdjustEnvironment(patch);
+        AddIceChunks(patch);
+        LogEvent(patch, totalTimePassed);
+    }
+
+    private void AdjustBackground(Patch patch)
+    {
+        patch.BiomeTemplate.Background = Background;
+        patch.CurrentSnapshot.Background = Background;
+        patch.BiomeTemplate.Sunlight.Colour = new Color(0.8f, 0.8f, 1, 1);
+        patch.CurrentSnapshot.LightColour = new Color(0.8f, 0.8f, 1, 1);
+    }
+
+    private void AdjustEnvironment(Patch patch)
+    {
+        var currentTemperature = patch.Biome.ChangeableCompounds[Compound.Temperature];
+        var currentSunlight = patch.Biome.ChangeableCompounds[Compound.Sunlight];
+
+        currentTemperature.Ambient = random.Next(-1, 5) - currentTemperature.Ambient;
+        currentSunlight.Ambient = 0.5f - currentSunlight.Ambient;
+
+        var changes = new Dictionary<Compound, float>
+        {
+            [Compound.Temperature] = currentTemperature.Ambient,
+            [Compound.Sunlight] = currentSunlight.Ambient,
+        };
+
+        patch.Biome.ApplyLongTermCompoundChanges(patch.BiomeTemplate, changes, new Dictionary<Compound, float>());
```

Comment:
As this method is applied to multiple patches the changes and the blank created dictionary here should both be class fields so that they don't need to be re-allocated for each method call.